### PR TITLE
DBZ-2329 Fix for: Change events lost when connnector is restarted while processing transaction with PK update

### PR DESCRIPTION
**Pull request by korzenek**
https://issues.redhat.com/browse/DBZ-2329

### CODE REVIEW

**Comment by jpechane** [comment]
@korzenek Hi, I think the solution is definitely interesting. But I am unsure about its scope. When this situation can happen - only for tx and change LSN that was stored in the offsets? If it is the case then I propose to check the monotny failure only for that coordinates and then when any of the LSNs changes I'd return `changesStoppedBeingMonotonic` back to `false. Do you think this is feasible?

**Comment by korzenek** [comment]
@jpechane Hello
This situation can happen for all transactions which were aborted in the middle with offset commit.
Change_lsn's can and are different like in example I've attached to Jira issue.
From what I've observed, it seems, that changes can be not monotonic only in within transaction (common commit_lsn), but we can't be sure as I can't find this being documented in sql server docs.

`changesStoppedBeingMonotonic` variable is only used when checking if we should skip some entries after connector restart, which is based on comparison with `lastProcessedPositionOnStart`
We can set `changesStoppedBeingMonotonic` back to false when commit_lsn changes. But i'm not sure if it is necessary.

**Comment by jpechane** [comment]
@korzenek So the problem applies only to a single TX as no more than one TX can be aborted in the middle. Then I'd still prefer to reset the flag to be really sure it applies to that only one case.

**Comment by korzenek** [comment]
@jpechane 
I've pushed change which resets value of `changesStoppedBeingMonotonic` back to false when TX changes

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java [hunk 1]
```diff
@@ -191,15 +193,28 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
                                 tableWithSmallestLsn.next();
                                 continue;
                             }
+
+                            if (tableWithSmallestLsn.isNewTransaction() && changesStoppedBeingMonotonic.get()) {
+                                LOGGER.info("Resetting changesStoppedBeingMonotonic as transaction changes");
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java:198]: That's more for DEBUG level perhaps? Or is this a rare thing to happen?

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java [hunk 2]
```diff
@@ -191,15 +193,28 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
...
+                                changesStoppedBeingMonotonic.set(false);
+                            }
+
+                            // After restart for changes that are not monotonic to avoid data loss
+                            if (tableWithSmallestLsn.isCurrentPositionSmallerThanPreviousPosition()) {
+                                LOGGER.info("Disabling skipping changes due to not monotonic order of changes");
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java:204]: Same here.

File: debezium-core/src/main/java/io/debezium/pipeline/source/spi/ChangeTableResultSet.java [hunk 3]
```diff
@@ -100,4 +110,6 @@ public String toString() {
     protected abstract int getOperation(ResultSet resultSet) throws SQLException;
 
     protected abstract T getNextChangePosition(ResultSet resultSet) throws SQLException;
+
+    protected abstract boolean isNewTransaction() throws SQLException;
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/source/spi/ChangeTableResultSet.java:114]: Could you add a JavaDoc comment describing the semantics; in particular, what's "new" in this context?

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java [hunk 4]
```diff
@@ -605,6 +605,56 @@ public void updatePrimaryKeyWithRestartInMiddle() throws Exception {
...
+
+        final SourceRecords records1 = consumeRecordsByTopic(14);
+
+        stopConnector();
+
+        start(SqlServerConnector.class, config);
+        assertConnectorIsRunning();
+        final SourceRecords records2 = consumeRecordsByTopic(6);
+
+        final List<SourceRecord> tableC = records1.recordsForTopic("server1.dbo.tableb");
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java:650]: tableC -> tableB?

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java (hunk 1 above)
Comment (jpechane) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java:198]: IIUC it should happen only once upon start then I'd keep it as info,

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java (hunk 2 above)
Comment (jpechane) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java:204]: Same as above

File: debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java (hunk 1 above)
Comment (gunnarmorling) [reply debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java:198]: Ok, makes sense.

File: debezium-core/src/main/java/io/debezium/pipeline/source/spi/ChangeTableResultSet.java (hunk 3 above)
Comment (korzenek) [reply debezium-core/src/main/java/io/debezium/pipeline/source/spi/ChangeTableResultSet.java:119]: Added.

**Comment by gunnarmorling** [comment]
Rebased and applied; thanks, @korzenek. The test failure should be unrelated; in fact I believe this was fixed already.

**Pull request merged**